File: src/scripts/import_chainlink_from_csv.py
import csv
import os
import psycopg2
from psycopg2.extras import execute_values

# Database connection parameters (these should be in environment variables in production)
DB_HOST = "localhost"  # Replace with your Supabase database host
DB_PORT = "5432"       # Default PostgreSQL port
DB_NAME = "postgres"   # Default database name
DB_USER = "postgres"   # Default user
DB_PASSWORD = ""       # Add your password here

# CSV file path
CSV_FILE_PATH = r"C:\Users\info\Downloads\Untitled spreadsheet - Sheet1.csv"

def connect_to_db():
    """Connect to the PostgreSQL database server"""
    conn = None
    try:
        # Connect to the PostgreSQL server
        print('Connecting to the PostgreSQL database...')
        conn = psycopg2.connect(
            host=DB_HOST,
            port=DB_PORT,
            database=DB_NAME,
            user=DB_USER,
            password=DB_PASSWORD
        )
        return conn
    except (Exception, psycopg2.DatabaseError) as error:
        print(f"Error connecting to the database: {error}")
        if conn is not None:
            conn.close()
        raise

def import_data_from_csv():
    """Import data from CSV file to the database"""
    conn = None
    try:
        conn = connect_to_db()
        cursor = conn.cursor()
        
        # First, truncate the existing table to avoid duplicates
        print("Truncating existing table...")
        cursor.execute("TRUNCATE TABLE public.chainlink_products;")
        
        # Read the CSV file and prepare data for insertion
        print(f"Reading data from {CSV_FILE_PATH}...")
        with open(CSV_FILE_PATH, 'r', encoding='utf-8') as csv_file:
            csv_reader = csv.DictReader(csv_file)
            products = []
            
            for row in csv_reader:
                # Convert empty strings to None for nullable fields
                for key, value in row.items():
                    if value == '':
                        row[key] = None
                
                # Handle numeric values
                try:
                    supply_price = float(row['supply_price']) if row['supply_price'] else 0
                except (ValueError, TypeError):
                    supply_price = 0
                
                try:
                    retail_price = float(row['retail_price']) if row['retail_price'] else 0
                except (ValueError, TypeError):
                    retail_price = 0
                
                # Create a tuple with the product data
                product = (
                    row['sku'],
                    row['name'],
                    row['product_category'],
                    row['variant_option_one_name'],
                    row['variant_option_one_value'],
                    row['variant_option_two_name'],
                    row['variant_option_two_value'],
                    row['variant_option_three_name'],
                    row['variant_option_three_value'],
                    supply_price,
                    retail_price
                )
                products.append(product)
        
        # Insert data in batches for better performance
        batch_size = 100
        total_products = len(products)
        print(f"Inserting {total_products} products into the database...")
        
        # SQL for inserting products
        insert_sql = """
            INSERT INTO public.chainlink_products (
                sku, name, product_category,
                variant_option_one_name, variant_option_one_value,
                variant_option_two_name, variant_option_two_value,
                variant_option_three_name, variant_option_three_value,
                supply_price, retail_price
            ) VALUES %s
        """
        
        # Insert data in batches
        for i in range(0, total_products, batch_size):
            batch = products[i:i + batch_size]
            execute_values(cursor, insert_sql, batch)
            conn.commit()
            print(f"Inserted batch {i // batch_size + 1} ({len(batch)} products)")
        
        # Verify the count in the database
        cursor.execute("SELECT COUNT(*) FROM public.chainlink_products;")
        count = cursor.fetchone()[0]
        print(f"Total products in database after import: {count}")
        
        # Close the database connection
        cursor.close()
        conn.close()
        print("Database connection closed.")
        
    except (Exception, psycopg2.DatabaseError) as error:
        print(f"Error: {error}")
        if conn is not None:
            conn.close()
            print("Database connection closed.")

if __name__ == "__main__":
    import_data_from_csv()

File: src/scripts/import_chainlink_from_csv.js
const fs = require('fs');
const { createClient } = require('@supabase/supabase-js');
const csv = require('csv-parser');

// Supabase configuration
const supabaseUrl = process.env.SUPABASE_URL || 'your-supabase-url';
const supabaseKey = process.env.SUPABASE_KEY || 'your-supabase-key';
const supabase = createClient(supabaseUrl, supabaseKey);

// CSV file path
const CSV_FILE_PATH = 'C:\\Users\\info\\Downloads\\Untitled spreadsheet - Sheet1.csv';

// Function to process the CSV file
async function importChainlinkProducts() {
  console.log('Starting import process...');
  
  try {
    // First, truncate the existing table
    console.log('Truncating existing table...');
    const { error: truncateError } = await supabase.rpc('truncate_chainlink_products');
    
    if (truncateError) {
      console.error('Error truncating table:', truncateError);
      return;
    }
    
    // Read and process the CSV file
    console.log(`Reading data from ${CSV_FILE_PATH}...`);
    const products = [];
    
    // Create a promise to read the CSV file
    const readCsvPromise = new Promise((resolve, reject) => {
      fs.createReadStream(CSV_FILE_PATH)
        .pipe(csv())
        .on('data', (row) => {
          // Convert empty strings to null
          Object.keys(row).forEach(key => {
            if (row[key] === '') {
              row[key] = null;
            }
          });
          
          // Handle numeric values
          const supplyPrice = row.supply_price ? parseFloat(row.supply_price) : 0;
          const retailPrice = row.retail_price ? parseFloat(row.retail_price) : 0;
          
          // Create a product object
          const product = {
            sku: row.sku,
            name: row.name,
            product_category: row.product_category,
            variant_option_one_name: row.variant_option_one_name,
            variant_option_one_value: row.variant_option_one_value,
            variant_option_two_name: row.variant_option_two_name,
            variant_option_two_value: row.variant_option_two_value,
            variant_option_three_name: row.variant_option_three_name,
            variant_option_three_value: row.variant_option_three_value,
            supply_price: supplyPrice,
            retail_price: retailPrice
          };
          
          products.push(product);
        })
        .on('end', () => {
          console.log(`Finished reading ${products.length} products from CSV`);
          resolve(products);
        })
        .on('error', (error) => {
          reject(error);
        });
    });
    
    // Wait for CSV reading to complete
    await readCsvPromise;
    
    // Insert data in batches for better performance
    const batchSize = 100;
    const totalProducts = products.length;
    console.log(`Inserting ${totalProducts} products into the database...`);
    
    // Process in batches
    for (let i = 0; i < totalProducts; i += batchSize) {
      const batch = products.slice(i, i + batchSize);
      const batchNumber = Math.floor(i / batchSize) + 1;
      
      console.log(`Inserting batch ${batchNumber} (${batch.length} products)`);
      
      const { error: insertError } = await supabase
        .from('chainlink_products')
        .insert(batch);
      
      if (insertError) {
        console.error(`Error inserting batch ${batchNumber}:`, insertError);
      } else {
        console.log(`Successfully inserted batch ${batchNumber}`);
      }
    }
    
    // Verify the count in the database
    const { data: countData, error: countError } = await supabase
      .from('chainlink_products')
      .select('*', { count: 'exact', head: true });
    
    if (countError) {
      console.error('Error getting count:', countError);
    } else {
      console.log(`Total products in database after import: ${countData.count}`);
    }
    
    console.log('Import process completed.');
    
  } catch (error) {
    console.error('Error during import process:', error);
  }
}

// Create the stored procedure for truncating the table (run this once)
async function createTruncateProcedure() {
  const { error } = await supabase.rpc('create_truncate_procedure');
  
  if (error) {
    console.error('Error creating procedure:', error);
  } else {
    console.log('Truncate procedure created successfully');
  }
}

// Function to create the stored procedure
async function setupDatabase() {
  // Create the stored procedure for truncating the table
  const { error } = await supabase.rpc('execute_sql', {
    sql_query: `
      CREATE OR REPLACE FUNCTION truncate_chainlink_products()
      RETURNS void AS $$
      BEGIN
        TRUNCATE TABLE public.chainlink_products;
      END;
      $$ LANGUAGE plpgsql SECURITY DEFINER;
    `
  });
  
  if (error) {
    console.error('Error setting up database:', error);
  } else {
    console.log('Database setup completed');
    // Now run the import
    await importChainlinkProducts();
  }
}

// Run the script
setupDatabase().catch(console.error);
